modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/PartitionPreloadingRoutine.java

```diff
@@ -0,0 +1,566 @@
+
+    /** Remaining nodes with groups and the number of partitions. */
+    @GridToStringInclude
+    private final Map<UUID, Map<Integer, Set<Integer>>> remaining;
+
+    /** Count of partition snapshots received. */
+    private final AtomicLong receivedCnt = new AtomicLong();
+
+    /** Cache group with restored partition snapshots and HWM value of update counter mapped to node identifier. */
+    @GridToStringInclude
+    private final Map<Integer, Map<UUID, Map<Integer, Long>>> restored = new ConcurrentHashMap<>();
+
+    /**
+     * Cache group identifiers with historical assignments future that will be completed when partition files are
+     * preloaded.
+     */
+    private final Map<Integer, GridFutureAdapter<GridDhtPreloaderAssignments>> futAssigns = new ConcurrentHashMap<>();
+
+    /** Total number of partitions. */
+    private final long totalPartitionsCnt;
+
+    /** Snapshot future. */
+    private IgniteInternalFuture<Boolean> snapshotFut;
+
+    /**
+     * @param exchFut Exchange future.
+     * @param cctx Cache shared context.
+     * @param rebalanceId Rebalance ID
+     * @param assignments Assignments mapped by node ID.
+     */
+    public PartitionPreloadingRoutine(
+        GridDhtPartitionsExchangeFuture exchFut,
+        GridCacheSharedContext cctx,
+        long rebalanceId,
+        Map<UUID, Map<Integer, Set<Integer>>> assignments
+    ) {
+        long totalParts = 0;
+
+        // Copy contents.
+        Map<UUID, Map<Integer, Set<Integer>>> remaining0 = U.newHashMap(assignments.size());
+
+        for (Map.Entry<UUID, Map<Integer, Set<Integer>>> nodeAssign : assignments.entrySet()) {
+            Map<Integer, Set<Integer>> nodeAssign0 = new ConcurrentHashMap<>(nodeAssign.getValue().size());
+
+            remaining0.put(nodeAssign.getKey(), nodeAssign0);
+
+            for (Map.Entry<Integer, Set<Integer>> grpAssign : nodeAssign.getValue().entrySet()) {
+                nodeAssign0.put(grpAssign.getKey(), new GridConcurrentHashSet<>(grpAssign.getValue()));
+                futAssigns.put(grpAssign.getKey(), new GridFutureAdapter<>());
+
+                totalParts += grpAssign.getValue().size();
+            }
+        }
+
+        this.cctx = cctx;
+        this.rebalanceId = rebalanceId;
+
+        exchId = exchFut.exchangeId();
+        topVer = exchFut.topologyVersion();
+        log = cctx.kernalContext().log(getClass());
+        totalPartitionsCnt = totalParts;
+        remaining = Collections.unmodifiableMap(remaining0);
+
+        restoreHnd = new PartitionRestoreHandler(cctx);
+    }
+
+    /**
+     * Start partitions preloading.
+     *
+     * @return Cache group identifiers with futures that will be completed when partitions are preloaded.
+     */
+    public Map<Integer, IgniteInternalFuture<GridDhtPreloaderAssignments>> startPartitionsPreloading() {
+        assert !remaining.isEmpty();
+
+        restoreHnd.start();
+
+        requestPartitionsSnapshot(remaining.entrySet().iterator());
```

Comment:
Please, use a more straight forward way instead of the self-iterable `requestPartitionsSnapshot` method. 

For instance,

```
        IgniteInternalFuture<?> reqSnapshotChain = new GridFinishedFuture<>();
        
        for (Map.Entry<UUID, Map<Integer, Set<Integer>>> e : remaining.entrySet())
            reqSnapshotChain = reqSnapshotChain.chain(f -> requestPartitionsSnapshot(e.getKey(), e.getValue()));
```